File: bindings/javascript/src/opcua.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// OPC UA protocol types for proven-servers.

/** ServiceType matching the Idris2 ABI tags. */
export const ServiceType = Object.freeze({
  READ: 0,
  WRITE: 1,
  BROWSE: 2,
  SUBSCRIBE: 3,
  PUBLISH: 4,
  CALL: 5,
  CREATE_SESSION: 6,
  ACTIVATE_SESSION: 7,
  CLOSE_SESSION: 8,
  CREATE_SUBSCRIPTION: 9,
  DELETE_SUBSCRIPTION: 10,
});

/** NodeClass matching the Idris2 ABI tags. */
export const NodeClass = Object.freeze({
  OBJECT: 0,
  VARIABLE: 1,
  METHOD: 2,
  OBJECT_TYPE: 3,
  VARIABLE_TYPE: 4,
  REFERENCE_TYPE: 5,
  DATA_TYPE: 6,
  VIEW: 7,
});

/** StatusCode matching the Idris2 ABI tags. */
export const StatusCode = Object.freeze({
  GOOD: 0,
  UNCERTAIN: 1,
  BAD: 2,
  BAD_NODE_ID_UNKNOWN: 3,
  BAD_ATTRIBUTE_ID_INVALID: 4,
  BAD_NOT_READABLE: 5,
  BAD_NOT_WRITABLE: 6,
  BAD_OUT_OF_RANGE: 7,
  BAD_TYPE_MISMATCH: 8,
  BAD_SESSION_ID_INVALID: 9,
  BAD_SUBSCRIPTION_ID_INVALID: 10,
  BAD_TIMEOUT: 11,
});

/** SecurityMode matching the Idris2 ABI tags. */
export const SecurityMode = Object.freeze({
  NONE: 0,
  SIGN: 1,
  SIGN_AND_ENCRYPT: 2,
});

/** SessionState matching the Idris2 ABI tags. */
export const SessionState = Object.freeze({
  IDLE: 0,
  CONNECTED: 1,
  CREATED: 2,
  ACTIVATED: 3,
  MONITORING: 4,
  CLOSING: 5,
});

File: bindings/dart/lib/src/opcua.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// OPC UA protocol types for proven-servers.

/// ServiceType matching the Idris2 ABI tags.
enum ServiceType {
  read(0),
  write(1),
  browse(2),
  subscribe(3),
  publish(4),
  call(5),
  createSession(6),
  activateSession(7),
  closeSession(8),
  createSubscription(9),
  deleteSubscription(10);

  const ServiceType(this.tag);
  final int tag;

  static ServiceType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// NodeClass matching the Idris2 ABI tags.
enum NodeClass {
  object(0),
  variable(1),
  method(2),
  objectType(3),
  variableType(4),
  referenceType(5),
  dataType(6),
  view(7);

  const NodeClass(this.tag);
  final int tag;

  static NodeClass? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// StatusCode matching the Idris2 ABI tags.
enum StatusCode {
  good(0),
  uncertain(1),
  bad(2),
  badNodeIdUnknown(3),
  badAttributeIdInvalid(4),
  badNotReadable(5),
  badNotWritable(6),
  badOutOfRange(7),
  badTypeMismatch(8),
  badSessionIdInvalid(9),
  badSubscriptionIdInvalid(10),
  badTimeout(11);

  const StatusCode(this.tag);
  final int tag;

  static StatusCode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// SecurityMode matching the Idris2 ABI tags.
enum SecurityMode {
  none(0),
  sign(1),
  signAndEncrypt(2);

  const SecurityMode(this.tag);
  final int tag;

  static SecurityMode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// SessionState matching the Idris2 ABI tags.
enum SessionState {
  idle(0),
  connected(1),
  created(2),
  activated(3),
  monitoring(4),
  closing(5);

  const SessionState(this.tag);
  final int tag;

  static SessionState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}
